### PR TITLE
Reduce overhead by changing alignment

### PR DESCRIPTION
**Pull request by DiazAlan**
Signed-off-by: AlanDiaz <diazalan@google.com>

Commit Message: Reduce overhead via alignment
Additional Description: Moved smaller values like enums, int32 and bools to the end of the class to save on memory via alignment
Risk Level: Low

### CODE REVIEW

**Comment by DiazAlan** [comment]
@adisuissa @yurykats

**Comment by yurykats** [comment]
LGTM. Out of curiosity, do you have sizeof number before and afer?

File: source/common/upstream/upstream_impl.h [hunk 1]
```diff
@@ -988,23 +985,29 @@ class ClusterInfoImpl : public ClusterInfo,
...
   mutable Http::Http2::CodecStats::AtomicPtr http2_codec_stats_;
   mutable Http::Http3::CodecStats::AtomicPtr http3_codec_stats_;
   UpstreamHttpFactoryContextImpl upstream_context_;
+
+  // Keep small values like bools and enums at the end of the class to reduce
+  // overhead via alignment
+  const uint32_t per_connection_buffer_limit_bytes_;
+  const uint32_t max_response_headers_count_;
+  LoadBalancerType lb_type_;
+  const bool drain_connections_on_host_removal_;
```
Comment (yurykats) [inline source/common/upstream/upstream_impl.h:1005]: May also pack booleans together with ":1"

File: source/common/upstream/upstream_impl.h [hunk 2]
```diff
@@ -988,23 +985,29 @@ class ClusterInfoImpl : public ClusterInfo,
...
   mutable Http::Http1::CodecStats::AtomicPtr http1_codec_stats_;
   mutable Http::Http2::CodecStats::AtomicPtr http2_codec_stats_;
   mutable Http::Http3::CodecStats::AtomicPtr http3_codec_stats_;
   UpstreamHttpFactoryContextImpl upstream_context_;
+
+  // Keep small values like bools and enums at the end of the class to reduce
+  // overhead via alignment
+  const uint32_t per_connection_buffer_limit_bytes_;
+  const uint32_t max_response_headers_count_;
+  LoadBalancerType lb_type_;
```
Comment (yurykats) [inline source/common/upstream/upstream_impl.h:1004]: May also shrink the enum to a single byte, by making it extend ` uint8_t` ( I did that in [#24644](https://github.com/envoyproxy/envoy/pull/24644/)

Comment (DiazAlan) [reply source/common/upstream/upstream_impl.h:1004]: Done

**Comment by DiazAlan** [comment]
@yurykats here's the size of comparison
New sizeof 2632
Old sizeof 2656

**Comment by yurykats** [comment]
LGTM, thanks!

**Review by yurykats — APPROVED** [review]

**Review by adisuissa — APPROVED** [review]
LGTM, thanks!

/assign-from @envoyproxy/senior-maintainers

File: source/common/upstream/upstream_impl.h [hunk 3]
```diff
@@ -988,23 +985,29 @@ class ClusterInfoImpl : public ClusterInfo,
...
+  const bool drain_connections_on_host_removal_ : 1;
+  const bool connection_pool_per_downstream_connection_ : 1;
+  const bool warm_hosts_ : 1;
+  const bool set_local_interface_name_on_upstream_connections_ : 1;
+  // true iff the cluster proto specified upstream http filters.
+  bool has_configured_http_filters_{false};
```
Comment (pradeepcrao) [inline source/common/upstream/upstream_impl.h:1010]: This needs to be 1 bit too.

Comment (DiazAlan) [reply source/common/upstream/upstream_impl.h]: Done. I found out you can't have a default value for a member and a bit field. So I moved the default value into the constructor

**Review by lizan — APPROVED** [review]

**Pull request merged**